### PR TITLE
[Merged by Bors] - feat(MeasureTheory/Function): Uniform tightness of functions in Lp

### PR DESCRIPTION
**Pull request by igorkhavkine**
Uniform tightness is a hypothesis necessary for the general version of the Vitali Convergence Theorem proven in #9163.

---

- [x] depends on: #14853

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/MeasureTheory/Function/UnifTight.lean
Comment (YaelDillies) [inline Mathlib/MeasureTheory/Function/UnifTight.lean:57]: I don't see the point of this sentence

File: Mathlib/MeasureTheory/Function/UnifTight.lean
Comment (YaelDillies) [inline Mathlib/MeasureTheory/Function/UnifTight.lean:77]:
```suggestion
if and only if it is uniformly integrable, uniformly tight and converges to `g` in measure. -/
```

File: Mathlib/MeasureTheory/Function/UnifTight.lean
Comment (YaelDillies) [inline Mathlib/MeasureTheory/Function/UnifTight.lean:82]: What's this sorry for?

File: Mathlib/MeasureTheory/Function/UnifTight.lean
Comment (YaelDillies) [inline Mathlib/MeasureTheory/Function/UnifTight.lean:31]: This lemma doesn't exist

File: Mathlib/MeasureTheory/Function/UnifTight.lean
Comment (YaelDillies) [inline Mathlib/MeasureTheory/Function/UnifTight.lean:29]: Neither does this one

**Comment by igorkhavkine** [comment]
Just to clarify, how self contained should this PR be (given that it should be subsumed by #9163)? I could just remove anything not directly pertaining to the UnifTight definition, including the formalization of the statement of the general Vitali Convergence Theorem (it's sorry-ed here because the proof takes up all of #9163).

**Comment by YaelDillies** [comment]
It should be self-contained and only pertain to uniform tightness

**Comment by igorkhavkine** [comment]
OK, I've made the PR as much self contained as I think it makes sense.

**Review by YaelDillies — COMMENTED** [review]
Sorry, I'm being picky here, but hopefully it teaches you how to best split a PR, which will make both your and the reviewers' lives better.

File: Mathlib/MeasureTheory/Function/UnifTight.lean
Comment (igorkhavkine) [reply Mathlib/MeasureTheory/Function/UnifTight.lean:42]: Sorry, I guess I really didn't understand what you meant by self-contained. Can you clarify whether the this PR should reproduce the contents of `section UnifTight` in the original PR, or just the smaller `namespace UnifTight`?

<probably uninformed opinion>It seems to me that we are getting dangerously close to creating pull requests for the sake of creating pull requests. Whatever logical structure can be created at the level of the hierarchy of PRs, it seems to me, is already present in the original UnifTight.lean source file, at the level of code sections and namespaces.</opinion>

Comment (YaelDillies) [reply Mathlib/MeasureTheory/Function/UnifTight.lean:42]: The problem with #9163 as it stands is that its PR title does not in any way indicate that it's introducing a definition of uniform tightness. So either:
* You make the title of #9163 be about uniform tightness
* You split off the definition of uniform tightness (and anything that is more relevant to uniform tightness than the extension of Vitali convergence) to a separate PR (this one).

Depending on how you title the PR, having everything in one PR implies either
* that uniform tightness is mathematically used exclusively as an input to the more general version of Vitali convergence
* that the more general version of Vitali convergence is API to the definition of uniform tightness

I definitely disagree with the first, and probably also disagrees with the second.

Comment (YaelDillies) [reply Mathlib/MeasureTheory/Function/UnifTight.lean:42]: Pragmatically, I would expect the current PR to contain as much as possible from `MeasureTheory.Function.UnifTight` in #9163 that doesn't require changes to other files. Hopefully that represents the bulk of the file, meaning that we can get most of it merged quickly while other dependencies of #9163 are getting reviewed. Then #9163 will contain not much more than the extension of Vitali convergence.

Comment (igorkhavkine) [reply Mathlib/MeasureTheory/Function/UnifTight.lean:42]: OK, thanks. That clarifies what you were after. I will try to do the split later today.

> Depending on how you title the PR, having everything in one PR implies either
> 
>     * that uniform tightness is mathematically used exclusively as an input to the more general version of Vitali convergence
> 
>     * that the more general version of Vitali convergence is API to the definition of uniform tightness
> 
> I definitely disagree with the first, and probably also disagrees with the second.

While you are certainly right at the level of mathematics, arguably, at the level of Mathlib *right now* either one of these possibilities is as good as true. Perhaps I'm being naive here, but it's hard for me to see how the PR hierarchy (which AFAIK disappears into the aether as soon as all the PRs are merged) is relevant to the variation in the domain of use of tightness vs Vitali in Mathlib *in the future*. If there's some point of principle here about structuring PRs, perhaps it could be explained better to beginners. //opinion

Comment (YaelDillies) [reply Mathlib/MeasureTheory/Function/UnifTight.lean:42]: Well, you definitely want the very few people who care about uniform tightness or Vitali convergence to comment on your PRs and having a PR title that advertises your PR as relevant is a great first step. Note that, even if **you** think topics A and B are very closed, there will be people who care about A but not B, or B but not A. I can't count anymore the number of PRs which I should have reviewed but missed because they had a misleading title/a title about other changes that were lumped with the ones I cared about.

On a more long-termy note, commits are part of the history of Mathlib forever, so it's definitely worth spending *some* time thinking about how to present things to future git archeologists. Eg early PRs like https://github.com/leanprover-community/mathlib/pull/327 are really hard to understand and contain much less documentation than they should have. Splitting PRs really helps here.

Comment (igorkhavkine) [reply Mathlib/MeasureTheory/Function/UnifTight.lean]: > Pragmatically, I would expect the current PR to contain as much as possible from `MeasureTheory.Function.UnifTight` in #9163 that doesn't require changes to other files. Hopefully that represents the bulk of the file, meaning that we can get most of it merged quickly while other dependencies of #9163 are getting reviewed. Then #9163 will contain not much more than the extension of Vitali convergence.

Alright, probably this version of the PR is closer to what you were expecting. It still spans multiple files, due to the rearrangements that Yury had done earlier.

If you find some code/comment formatting changes that you would like to make, just go ahead and suggest them. I'm quite indifferent about them and I doubt that I would find all of them myself.

File: Mathlib/MeasureTheory/Function/UnifTight.lean
Comment (YaelDillies) [inline Mathlib/MeasureTheory/Function/UnifTight.lean:212]: I would leave this section out until #9163

**Review by YaelDillies — COMMENTED** [review]
Yes, thanks. This looks much better. This PR now clearly works towards a specific goal (namely defining uniform tightness and proving basic properties)

File: Mathlib/MeasureTheory/Integral/Lebesgue.lean
Comment (YaelDillies) [inline Mathlib/MeasureTheory/Integral/Lebesgue.lean:1438]: Can you add the `setIntegral` version as well?

File: Mathlib/MeasureTheory/Function/UnifTight.lean
Comment (igorkhavkine) [reply Mathlib/MeasureTheory/Function/UnifTight.lean:53]: This was Yury's design decision (cf [this commit](https://github.com/leanprover-community/mathlib4/pull/9163/commits/d4b82684179c3fd3eeba5883fb1dfea98c240c19) and [comment](https://github.com/leanprover-community/mathlib4/pull/9163#issuecomment-1940535054)). I was originally following the analogy with `UnifIntegrable` that had `⦃ε : ℝ⦄`. Changing this will create some propagation.

Comment (YaelDillies) [reply Mathlib/MeasureTheory/Function/UnifTight.lean:53]: @urkud, can we have some background explanation?

File: Mathlib/MeasureTheory/Integral/Lebesgue.lean
Comment (igorkhavkine) [reply Mathlib/MeasureTheory/Integral/Lebesgue.lean:1438]: Done (now in `MeasureTheory.Integral.SetIntegral`). After a cursory search, I found only [one place](https://github.com/leanprover-community/mathlib4/blob/d419680bd7d6378744b29e71e3e161a996e82178/Mathlib/Probability/Martingale/Convergence.lean#L404) in Mathlib where `setIntegral_compl` could be usefully dropped in.

File: Mathlib/MeasureTheory/Function/UnifTight.lean
Comment (urkud) [reply Mathlib/MeasureTheory/Function/UnifTight.lean:53]: I think that we should have both versions. I don't care which one is the definition and which one is a lemma. Probably, I was wrong and the `NNReal` version should be the definition.

Comment (YaelDillies) [reply Mathlib/MeasureTheory/Function/UnifTight.lean:53]: I think that's a good idea

Comment (igorkhavkine) [reply Mathlib/MeasureTheory/Function/UnifTight.lean:53]: > I think that's a good idea

I don't know. This sounds like a doubling of the API in size, for both versions of the uniform tightness condition. Is there any example in Mathlib where something like that is actually done (in an economical way, if possible)?

Comment (YaelDillies) [reply Mathlib/MeasureTheory/Function/UnifTight.lean:53]: Why would it double the size of the API? All it's going to add is a lemma saying that the definitions in terms of `NNReal` and `ENNReal` agree.

Comment (igorkhavkine) [reply Mathlib/MeasureTheory/Function/UnifTight.lean:53]: Sorry, I'm probably missing the point again. If there are multiple versions of `UnifTight` (`\eps` being `Real`, `NNReal` or `ENNReal`), then what is the version to be used in any lemma/theorem that relies on `UnifTight`? Naively, one would add all the corresponding versions, increasing the API by a factor of 2 or 3.

If no that, then one version of `UnifTight` should be primary and only that one used in the API. So which version should that be?

Comment (YaelDillies) [reply Mathlib/MeasureTheory/Function/UnifTight.lean:53]: When several definitions A and B are equivalent, you don't write `def A`, `def B`, `lemma _ : A \iff B` but `def A`, `lemma _ : A \iff B`.

Here the definition should be in terms of `NNReal` and you can have a lemma stating the equivalence with the `ENNReal` version.

Comment (igorkhavkine) [reply Mathlib/MeasureTheory/Function/UnifTight.lean:53]: OK, one last comment before the decision is made. The type of `ε` already came up already early on (cf [this comment](https://github.com/leanprover-community/mathlib4/pull/9163/files#r1493852124)). The one argument I can see in favor of `ε : ENNReal` is that in all the proofs `ε` is always compared with `ENNReal` because that is the target type of [docs#MeasureTheory.snorm](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Function/LpSeminorm/Basic.html#MeasureTheory.snorm), which makes some internal manipulations easier.

The tradeoffs are:
* `ε : Real`: mathematically natural, annoying casts in proofs
* `ε : NNReal`: still mathematically natural, still some (but less annoying) casts in proofs
* `ε : ENNReal`: always check `∞` as a special case (unnatural?), no casts in proofs

So, final decision?

Comment (YaelDillies) [reply Mathlib/MeasureTheory/Function/UnifTight.lean:53]: I already knew this tradeoff, so I'm happy with `NNReal` 😁

File: Mathlib/MeasureTheory/Function/LpSeminorm/Basic.lean
Comment (YaelDillies) [inline Mathlib/MeasureTheory/Function/LpSeminorm/Basic.lean:1174]: Why did you change `ENNReal` to `NNReal` here? It was more general before (even if trivially)

File: Mathlib/MeasureTheory/Function/UnifTight.lean
Comment (YaelDillies) [inline Mathlib/MeasureTheory/Function/UnifTight.lean:55]:
```suggestion
theorem unifTight_iff_ennreal {_ : MeasurableSpace α} (f : ι → α → β) (p : ℝ≥0∞) (μ : Measure α) :
```
looks more accurate

File: Mathlib/MeasureTheory/Function/UnifTight.lean
Comment (YaelDillies) [inline Mathlib/MeasureTheory/Function/UnifTight.lean:71]: and similarly 
```suggestion
theorem unifTight_iff_real {_ : MeasurableSpace α} (f : ι → α → β) (p : ℝ≥0∞) (μ : Measure α) :
```

File: Mathlib/MeasureTheory/Function/UnifTight.lean
Comment (YaelDillies) [inline Mathlib/MeasureTheory/Function/UnifTight.lean:144]:
```suggestion
/-- If two functions agree a.e., then one is tight iff the other is tight. -/
```
just for clarity

File: Mathlib/MeasureTheory/Function/LpSeminorm/Basic.lean
Comment (igorkhavkine) [reply Mathlib/MeasureTheory/Function/LpSeminorm/Basic.lean:1174]: ¯\_(ツ)_/¯ I guess with that philosophy, there was no reason to change `UnifTight` to use `ENNReal` in the first place. I changed it here because it made some of the proofs I was reworking easier. Right now this doesn't contradict any convention for what type `ε` and `δ` in `Analysis` or `MeasureTheory`, since I don't see any unified convention.

Comment (YaelDillies) [reply Mathlib/MeasureTheory/Function/LpSeminorm/Basic.lean:1174]: > I guess with that philosophy, there was no reason to change `UnifTight` to use `ENNReal` in the first place

I mean... yes, I agree? But that doesn't go against my comment above: Have easy to prove definitions, hard to prove theorems.

Comment (igorkhavkine) [reply Mathlib/MeasureTheory/Function/LpSeminorm/Basic.lean:1174]: OK, I can understand that. Apologies for being dense, but as I'm still learning about different Mathlib philosophies and conventions it'd be nice to hear some general guidelines or explanations along with patch change suggestions. Understanding some general guidelines better I think would also help to smoothly update the parent PR.

Comment (YaelDillies) [reply Mathlib/MeasureTheory/Function/LpSeminorm/Basic.lean]: That's why I keep on commenting here 😃

There's a lot of unwritten community knowledge that you will learn by just asking. We all wish it were written down, but nobody has done that yet. So please keep on asking questions.

File: Mathlib/MeasureTheory/Function/UnifTight.lean
Comment (YaelDillies) [inline Mathlib/MeasureTheory/Function/UnifTight.lean:58]:
```suggestion
  simp [ENNReal.forall_ennreal, UnifTight]
```

File: Mathlib/MeasureTheory/Function/UnifTight.lean
Comment (YaelDillies) [inline Mathlib/MeasureTheory/Function/UnifTight.lean:214]:
```suggestion
  simpa [g] using hfε (hn.some i)
```
or something like that

File: Mathlib/MeasureTheory/Integral/Lebesgue.lean
Comment (YaelDillies) [reply Mathlib/MeasureTheory/Integral/Lebesgue.lean:1438]: Yeah, looks like a good idea to use it there. Mind opening a separate PR for `setIntegral_compl`/`setLintegral_compl` and the potential use you found?

**Comment by YaelDillies** [comment]
Can you merge master to reduce the diff?

**Comment by igorkhavkine** [comment]
> Can you merge master to reduce the diff?

Done. The build is marked as failed because of a linter complaint, which I'm not sure what to do about.

Also, I was undecided about choosing a proof for `unifTight_iff_real`. Please have a look.

**Comment by YaelDillies** [comment]
> The build is marked as failed because of a linter complaint, which I'm not sure what to do about.

You need to run `lake exe lint_style --update`

> I was undecided about choosing a proof for `unifTight_iff_real`.

I like the second one better. Also note that you can(/should?) incorporate the `exact` into the `refine`.

**Comment by igorkhavkine** [comment]
> > The build is marked as failed because of a linter complaint, which I'm not sure what to do about.
> 
> You need to run `lake exe lint_style --update`

That reproduces the same error locally, I think. Anyway the linter seems to be suggesting splitting MeasureTheory/Integral/Lebesgue.lean because it has grown too large. This patch adds about 50 lines to that file, which could potentially be moved elsewhere, but the file would still remain large. If a similar amount of code would be added to it on another occasion, the same error would come up. Perhaps it requires some higher level decisions about potentially splitting the file in a fine grained way. I don't feel qualified to decide that.

**Comment by YaelDillies** [comment]
@grunweg, what are we doing wrong? Shouldn't `lake exe lint_style --update` update the lint style file?

**Comment by grunweg** [comment]
> @grunweg, what are we doing wrong? Shouldn't `lake exe lint_style --update` update the lint style file?

That is implemented in #14697, which needs to be reviewed :-) Right now, the `--update` option does not quite do that yet. (I'm silently changing the behaviour of --update in that PR, on the grounds that the current behaviour is not what you want.)

**Comment by YaelDillies** [comment]
@igorkhavkine, can you run `scripts/update-style-exceptions.sh`

**Comment by grunweg** [comment]
That last commit was unexpected... I just pushed a fix to the style exceptions file.

**Review by YaelDillies — APPROVED** [review]
This looks good to me but I need a measure theorist to sign off.

maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**
